connector/src/main/java/com/google/cloud/spark/bigquery/v2/IntermediateDataCleaner.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.spark.bigquery.v2;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Responsible for recursively deleting the intermediate path. Implementing Runnable in order to act
```

Comment:
Not sure if this is a good idea to run data cleanup as a shutdown hook. Generally shutdown hooks should run quickly and it's entirely possible that the GCS clients and associated thread pools may be in the process of shutting down, so it's hard to guarantee that this cleanup actually happened.

Should we consider alternatives? For instance, can we upload the temporary files with random prefixes and use GCS lifecycle policies to automatically clean them up after a safe window?

Reply:
This is done just as a safe-guard in case `BigQueryIndirectDataSourceWriter.cleanTemporaryGcsPathIfNeeded` has not been called by an error. Relying on GCS object lifecycle will require the bucket to allow lifecycle management, which is not guaranteed.